### PR TITLE
Complete minor TODOs and tasks

### PR DESCRIPTION
**Pull request by skejserjensen**
This PR completes a collection of minor TODOs and tasks with approximately one TODO or task per commit.

### CODE REVIEW

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: server/Cargo.toml [hunk 1]
```diff
@@ -6,44 +6,30 @@ edition = "2021"
...
-prost = "0.11.3"
-libc = "0.2.138"
 serial_test = "0.9.0"
 sysinfo = "0.26.8"
+tempfile = "3.3.0"
 
-[[test]]
-name = "integration"
-path = "tests/integration_test.rs"
+# Make machete ignore all dependencies that are required but not explicitly used in the source code
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/Cargo.toml:33]: Fixed.

File: server/Cargo.toml [hunk 2]
```diff
@@ -6,44 +6,30 @@ edition = "2021"
...
-sqlparser = "0.28.0"
-bytes = "1.3.0"
+parking_lot = "0.12.1"
+rusqlite = { version = "0.28.0", features = ["bundled"] }
+snmalloc-rs = "0.3.3"
+sqlparser = "0.27.0"
+tokio = { version = "1.22.0", features = ["rt-multi-thread", "signal"] }
+tonic = "0.8.3"
 
 # Log is a dependency so the compile time filters for log and tracing can be set to the same value
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/Cargo.toml:22]: Fixed.

File: server/src/remote.rs [hunk 3]
```diff
@@ -175,17 +175,27 @@ impl FlightServiceHandler {
...
-        );
+        schema: &SchemaRef,
+        flight_data_stream: &mut Streaming<FlightData>,
+    ) -> Result<(), Status> {
+        // Retrieve the data until the request does not contain any more data.
+        while let Some(flight_data) = flight_data_stream.next().await {
+            let record_batch = self.flight_data_to_record_batch(&flight_data?, schema)?;
+            let storage_engine = self.context.storage_engine.write().await;
+
+            // Write record_batch to the table with table_name as a compressed Apache Parquet files.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/remote.rs:189]: Fixed.

File: server/src/storage/compressed_data_buffer.rs [hunk 4]
```diff
@@ -87,12 +87,14 @@ impl CompressedDataBuffer {
...
-    fn get_size_of_compressed_segments(compressed_segments: &RecordBatch) -> usize {
+    fn size_of_compressed_segments(compressed_segments: &RecordBatch) -> usize {
         let mut total_size: usize = 0;
 
         // Compute the total number of bytes of memory used by the columns.
         for column in compressed_segments.columns() {
-            // TODO: How is this calculated internally?
+            // Recursively compute the total number of bytes of memory used by a single column. It
+            // is the size of the types, e.g., Array, ArrayData, Buffer, and Bitmap, and the data in
+            // the Apache Arrow format as the buffers with the data and the null bitmap if it exists.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_buffer.rs:97]: I have tried to rewrite that part of the comment and added a link to the [Apache Arrow Columnar Format specification ](Apache Arrow Columnar Format specification )which provides information about how each type of Apache Arrow Array is organized in memory.

File: server/src/storage/data_transfer.rs [hunk 5]
```diff
@@ -481,12 +485,15 @@ mod tests {
         let path = local_data_folder_path.join(format!("compressed/{}", TABLE_NAME));
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/data_transfer.rs:485]: Fixed.

File: server/src/storage/data_transfer.rs [hunk 6]
```diff
@@ -357,15 +358,17 @@ mod tests {
     #[tokio::test]
     async fn test_add_non_existent_file() {
         let temp_dir = tempfile::tempdir().unwrap();
-        let path = temp_dir.path().join(format!("{}/compressed", TABLE_NAME));
+        let path = temp_dir
+            .path()
+            .join(format!("{}/{}", TABLE_NAME, COMPRESSED_DATA_FOLDER));
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/data_transfer.rs:363]: Fixed.

File: server/src/storage/uncompressed_data_buffer.rs [hunk 7]
```diff
@@ -233,25 +244,30 @@ impl fmt::Debug for UncompressedOnDiskDataBuffer {
     }
 }
 
-// TODO: Delete the spilled Apache Parquet file after it has been read?
 #[async_trait]
 impl UncompressedDataBuffer for UncompressedOnDiskDataBuffer {
     /// Read the data from the Apache Parquet file and return it as a [`RecordBatch`].
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/uncompressed_data_buffer.rs:249]: Fixed and documented when `ParquetError` is returned.

File: server/src/storage/mod.rs [hunk 8]
```diff
@@ -59,6 +58,12 @@ use crate::types::{Timestamp, TimestampArray, Value, ValueArray};
 
 // TODO: Look into custom errors for all errors in storage engine.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/mod.rs:59]: Removed the TODO and added a task to determine what the Rust community recommends with regards to error handling and to determine if we can remove the last few uses of [panic!(), e.g., by rejecting data points when the storage engine runs out of memory](https://github.com/ModelarData/ModelarDB-RS/blob/dev/fix-todos/server/src/storage/uncompressed_data_manager.rs#L388), so we can make our error management more consistent and ensure the server keep ingesting data no matter what error occurs.

File: server/src/storage/mod.rs [hunk 9]
```diff
@@ -136,8 +142,20 @@ impl StorageEngine {
         })
     }
 
-    /// Pass `data_points` to [`UncompressedDataManager`]. Return [`Ok`] if all of the data points
+    /// Pass `record_batch` to [`CompressedDataManager`]. Return [`Ok`] if all of the data points
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/mod.rs:145]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 10]
```diff
@@ -54,20 +56,54 @@ pub(super) struct CompressedDataManager {
...
             local_data_folder,
             compressed_data_buffers: HashMap::new(),
             compressed_queue: VecDeque::new(),
             compressed_remaining_memory_in_bytes: compressed_reserved_memory_in_bytes as isize,
             compressed_schema,
-        }
+        })
+    }
+
+    /// Write `record_batch` to the table with `table_name` as a compressed Apache Parquet files.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:80]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 11]
```diff
@@ -122,13 +158,21 @@ impl CompressedDataManager {
...
-    /// [`Vec`] is returned, while a [`DataRetrievalError`](ModelarDbError::DataRetrievalError) is
-    /// returned if the compressed files could not be listed.
+    /// table with `table_name` and contains compressed segments within the given range of time and
+    /// value. If some compressed data that belongs to `table_name` is still in memory, save it to
+    /// disk first. If no files belongs to the table with `table_name` an empty [`Vec`] is returned,
+    /// while a [`DataRetrievalError`](ModelarDbError::DataRetrievalError) is returned if:
+    /// * A table with `table_name` does not exist.
+    /// * The compressed files could not be listed.
+    /// * The end time is before the start time.
+    /// * The max value is larger than the min value.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:168]: Replaced "larger" with "smaller" to match "before".

File: server/src/storage/compressed_data_manager.rs [hunk 12]
```diff
@@ -229,12 +300,44 @@ impl CompressedDataManager {
     }
 }
 
+/// Return the [`ObjectMeta`] in `object_meta` if it represents an Apache Parquet file and if the
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:303]: Replaced "in" with "passed".

File: server/src/storage/compressed_data_manager.rs [hunk 13]
```diff
@@ -229,12 +300,44 @@ impl CompressedDataManager {
...
+/// timestamps and values in its filename overlaps with the time range given by `start_time` and
+/// `end_time` and the value range given by `min_value` and `max_value`, otherwise [`None`] is
+/// returned. Assumes the name of Apache Parquet files represented by `object_meta` has the
+/// following format: `start_timestamp-end_timestamp-min_value-max_value.parquet`, where both
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:307]: Replaced "-" with "_" and "_" with "-" for in the format description for both `is_object_meta_relevant()` and `is_compressed_file_within_time_and_value_range()`.

File: server/src/storage/uncompressed_data_manager.rs [hunk 14]
```diff
@@ -59,22 +62,51 @@ pub(super) struct UncompressedDataManager {
...
+        // Ensure the folder required by the uncompressed data manager exists.
+        let local_uncompressed_data_folder = local_data_folder.join(UNCOMPRESSED_DATA_FOLDER);
+        fs::create_dir_all(&local_uncompressed_data_folder)?;
+
+        // Add references to the uncompressed data buffers currently on disk.
+        let mut finished_data_buffers: VecDeque<Box<dyn UncompressedDataBuffer>> = VecDeque::new();
+        for maybe_folder_dir_entry in local_uncompressed_data_folder.read_dir()? {
+            let folder_dir_entry = maybe_folder_dir_entry?;
+
+            // unwrap() is safe as the file_name is a univariate id.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/uncompressed_data_manager.rs:85]: Fixed.

File: server/tests/integration_test.rs [hunk 15]
```diff
@@ -1,126 +1,157 @@
...
+/// they cannot be imported to this file.
 pub type ArrowValue = datafusion::arrow::datatypes::Float32Type;
 pub type ValueArray = PrimitiveArray<ArrowValue>;
 pub type TimeSeriesId = std::primitive::u64;
 pub type ArrowTimeSeriesId = datafusion::arrow::datatypes::UInt64Type;
 pub type ArrowTimestamp = TimestampMillisecondType;
 
-const TABLE_NAME: &str = "data";
-const ADDRESS: &str = "127.0.0.1";
+/// The different types of tables used in integration the tests.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/tests/integration_test.rs:40]: Fixed.

File: server/tests/integration_test.rs [hunk 16]
```diff
@@ -325,97 +429,95 @@ fn get_binary_directory() -> path::PathBuf {
...
+        let flight_service_client = FlightServiceClient::connect(address).await?;
+        Ok(flight_service_client)
     })
 }
 
-/// Create a normal table or model table with/without tags in the
-/// ModelarDB server with Arrow Flight using the `do_action()` SQL Parser.
-///
-/// mode options: NormalTable/ModelTable/ModelTableNoTag
+/// Create a normal table or model table with or without tags in the ModelarDB server through Apache
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/tests/integration_test.rs:476]: Fixed.

File: server/tests/integration_test.rs [hunk 17]
```diff
@@ -447,75 +549,86 @@ fn generate_random_message(tag: Option<&str>) -> RecordBatch {
     }
 }
 
-/// Creates and returns FlightData based on the messages inserted.
-fn create_flight_data_from_messages(messages: &Vec<RecordBatch>) -> Vec<FlightData> {
-    let flight_descriptor = FlightDescriptor::new_path(vec![TABLE_NAME.to_string()]);
+/// Create and returns [`FlightData`] based on the data_points inserted.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/tests/integration_test.rs:552]: Fixed.

File: server/tests/integration_test.rs [hunk 18]
```diff
@@ -447,75 +549,86 @@ fn generate_random_message(tag: Option<&str>) -> RecordBatch {
...
+    runtime.block_on(async {
+        flight_service_client
+            .do_action(request)
+            .await
+            .expect("Could not flush data.");
     })
 }
 
-/// Execute a query on the ModelarDB server through the `do_get()` endpoint and return it.
+/// Execute a query against the ModelarDB server through the `do_get()` endpoint and return it.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/tests/integration_test.rs:605]: Replaced all instances of "endpoint(s)" with "method(s)" in `integration_test.rs` as "endpoint(s)" was only used in `integration_test.rs`.

File: server/src/storage/data_transfer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/storage/data_transfer.rs]: Is it necessary to have the `storage::` prefix. It does not seem like other uses of the constant has it.

[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/data_transfer.rs]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 19]
```diff
@@ -335,8 +335,8 @@ async fn is_object_meta_relevant(
 /// Return [`true`] if the timestamps and values in `file name` overlaps with the time range given
 /// by `start_time` and `end_time` and the value range given by `min_value` and `max_value`,
 /// otherwise [`false`]. Assumes `file_name` has the following format:
-/// `start_timestamp-end_timestamp-min_value-max_value.parquet`, where both `start_timestamp` and
-/// `end_timestamp` are of the same unit as `start_time` and `end_time`.
+/// `start-timestamp_end-timestamp_min-value_max_value.parquet`, where both `start-timestamp` and
```
Comment (CGodiksen) [inline server/src/storage/compressed_data_manager.rs:338]: Very minor but it seems that `max_value` was not changed to `max-value`.

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: server/src/storage/compressed_data_manager.rs (hunk 19 above)
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:338]: Fixed.

**Review by CGodiksen — APPROVED** [review]

File: server/tests/integration_test.rs [hunk 20]
```diff
@@ -131,187 +162,260 @@ fn test_get_schema() {
...
-
-    let message = generate_random_message(None);
-    let flight_data_vec = create_flight_data_from_messages(&vec![message.clone()]);
+fn test_can_ingest_data_point_without_tags() {
+    let temp_dir = tempfile::tempdir().expect("Could not create a directory.");
+    let flight_server = start_modelardbd(temp_dir.path());
+
+    let runtime = Runtime::new().expect("Unable to initialize runtime.");
+    let mut flight_service_client =
+        create_apache_arrow_flight_service_client(&runtime, &HOST, 9999)
```
Comment (chrthomsen) [inline server/tests/integration_test.rs:219]: Why not also use &PORT set to 9999?

File: server/src/remote.rs [hunk 21]
```diff
@@ -175,17 +175,27 @@ impl FlightServiceHandler {
...
+        while let Some(flight_data) = flight_data_stream.next().await {
+            let record_batch = self.flight_data_to_record_batch(&flight_data?, schema)?;
+            let storage_engine = self.context.storage_engine.write().await;
+
+            // Write record_batch to the table with table_name as a compressed Apache Parquet file.
+            storage_engine
+                .insert_record_batch(table_name, record_batch)
+                .await
+                .map_err(|error| {
+                    Status::internal(format!("Data could not be ingested: {}", error))
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/remote.rs:194]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 22]
```diff
@@ -122,13 +158,21 @@ impl CompressedDataManager {
     }
 
     /// Return an [`ObjectMeta`] for each compressed file in `query_data_folder` that belongs to the
-    /// table with `table_name`. If some compressed data that belongs to `table_name` is still in
-    /// memory, save it to disk first. If no files belongs to the table with `table_name` an empty
-    /// [`Vec`] is returned, while a [`DataRetrievalError`](ModelarDbError::DataRetrievalError) is
-    /// returned if the compressed files could not be listed.
+    /// table with `table_name` and contains compressed segments within the given range of time and
+    /// value. If some compressed data that belongs to `table_name` is still in memory, save it to
+    /// disk first. If no files belongs to the table with `table_name` an empty [`Vec`] is returned,
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:163]: Fixed.

File: server/src/storage/compressed_data_manager.rs [hunk 23]
```diff
@@ -229,12 +300,44 @@ impl CompressedDataManager {
     }
 }
 
+/// Return the [`ObjectMeta`] passed as `object_meta` if it represents an Apache Parquet file and if
+/// the timestamps and values in its filename overlaps with the time range given by `start_time` and
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/compressed_data_manager.rs:304]: Fixed.

File: server/src/storage/mod.rs [hunk 24]
```diff
@@ -175,7 +189,7 @@ impl StorageEngine {
         let compressed_buffers = self.uncompressed_data_manager.flush().await;
         let hash_to_table_name = self
             .metadata_manager
-            .get_mapping_from_hash_to_table_name()
+            .mapping_from_hash_to_table_name()
             .map_err(|error| error.to_string())?;
 
         let tag_hash_one_bits: u64 = 18446744073709550592;
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/storage/mod.rs:195]: Fixed.

File: server/tests/integration_test.rs (hunk 20 above)
Comment (skejserjensen) [reply server/tests/integration_test.rs:219]: Fixed.

**Review by chrthomsen — APPROVED** [review]

**Pull request merged**